### ISSUE TITLE
FaH client keeps losing CUDA and/or OpenCL install info after 1-3 WUs

### ISSUE THREAD

**Issue by Inkertus** [FoldingAtHome/fah-client-bastet#301]
Freshly installed FaH client (latest standard release) initially finds and successfully runs WUs for the 4070 Super, then hours later after 1-3 WUs loses the OpenCL, CUDA, or both info and claims platform not found or a lib .so file error, despite running fine hours earlier. I'm wondering if distributing a Flatpak, Appimage, or Snap is a viable option to enhance compatibility going forward.

Zorin OS 17.2 (Ubuntu derivative with customizations)
Nvidia driver 550.

00:49:46:I3:Loaded 1 wus.
00:49:46:W :OpenCL not supported: clGetPlatformIDs() returned -1001
00:49:46:W :CUDA not supported: cuInit() returned 100
00:49:46:I3:gpus = {
00:49:46:I3: "gpu:07:00:00": {"vendor": 4318, "device": 10115, "type": "nvidia", "supported": false, "description": "AD104 [GeForce RTX 4070 SUPER]"}
00:49:46:I3:}

14:38:28:I1:WU618: Global context and integrator variables write interval: disabled
14:38:28:I1:WU618:There are 3 platforms available.
14:38:28:I1:WU618:Platform 0: Reference
14:38:28:I1:WU618:Platform 1: CPU
14:38:28:I1:WU618:Platform 2: CUDA
14:38:28:I1:WU618: cuda-device 0 specified
14:38:28:I1:WU618:opencl-device was set but OpenCL platform could not be found.
--keeps throwing bad work units endlessly

**Comment by muziqaz** [comment]
This does not look like fahclient fault. We have a lot of people running Ubuntu on nVidia on v8, and they have no issues, once they set up the drivers. There must be something going on OS level, that hides CUDA and OCL. Fahclient sees only what is exposed to it by OS.
Zorin as well as other Ubuntu/Debian based distros seem to act really funky with nVidia drivers :/
I always thought that AMD GPU set up in Linux for FAH was biggest pain, but at least for AMD once you go through initial pains to get things working, things just work on variety of distros, on v8 and v7

**Comment by Inkertus** [comment]
Do you think Apparmor could be in play here? I can't think of much else that would affect it.

**Comment by jhitze** [comment]
@Inkertus I actually just ran into this issue too. First symptom for me was it not getting any new WUs. So I restarted the client and it complained like yours that cuda didn't exist. I rebooted the computer and everything came back up normally. /shrug

**Comment by jhitze** [comment]
@muziqaz @jcoffland  So, I just found out this bastet/fahcore/CUDA bug killed my bonus status.

<img width="449" alt="Screenshot 2024-11-05 at 5 21 58 PM" src="https://github.com/user-attachments/assets/9e03cae9-c640-47e6-b137-071fad298f40">

# Logs
I got all the logs for the past few weeks in this [shared folder](https://purdue0-my.sharepoint.com/:f:/g/personal/jhitze_purdue_edu/EtChls8Z8uRKmIR5-svfsW4Bnu-DbFeVQjulQDGA80vWOg?e=zlNZpY)

# Computer

## host
dell laptop running xen-ng 8.3
AMD Ryzen 5 5600H (12 threads)
GA107M [GeForce RTX 3050 Mobile]

## VM
11 CPUs
Ubuntu 22.04.5 LTS
GPU Passthrough - GeForce RTX 3050 Mobile
10gb ram
10gb vhd

### nvidia-smi
<img width="732" alt="Screenshot 2024-11-05 at 5 01 08 PM" src="https://github.com/user-attachments/assets/8d03fffb-226e-43b8-b23e-94f1857c8dc5">

# Story
Starting in the Oct 31st log file, you can see it went from a few hundred kb per file to 64mb and 93mb.

From what I can see, it gets stuck in this loop:
* retrieve WU for GPU
* start GPU fahcore
* fahcore tried to initiate cuda context
* fahcore fails
* WU returned as `BAD_WORK_UNIT`
* WU credited

Repeat like 10k times in two days and then my account loses its bonus. 😢 

It was working stable for several days beforehand, and since disabling the GPU, it is stable again.
<img width="809" alt="Screenshot 2024-11-05 at 5 15 11 PM" src="https://github.com/user-attachments/assets/4fb52e56-3ced-4a49-b550-115a3a9c26b3">

I think this necessitates two things (in addition to fixing the vanishing CUDA):
1. A circuit breaker on the client so if it has 10 `BAD_WORK_UNIT` or other errors, it stops until a human intervenes.
2. `BAD_WORK_UNIT` should be at least a warning in the logs, and probably an error.

What do you think?

**Comment by muziqaz** [comment]
Both suggestions have been suggested already, and they are essential to be implemented as soon as possible, as you are not the only one who is failing WUs in thousands.
In regards to CUDA disappearing, this might be VM thing. There is no other way to explain this, unless nVidia drivers went completely bonkers. FAH sees what is exposed to it by the system.
All I can suggest is to do some common things I would do if things were not working for me in Linux:
add your username to video and render groups with:
sudo usermod -a -G render yourusername
sudo usermod -a -G video yourusername
plus:
sudo apt install ocl-icd-opencl-dev
Restart the VM.

However this might not work, since your stuff works, and then suddenly it doesn't, which is weird.
It would be interesting to see clinfo output when things are working vs when things are not working

**Comment by jhitze** [comment]
I did the CLInfo before rebooting the VM. I did a full restart of the machine, then did the clinfo again. I did a compare of the two files and there were no differences.

[clinfo-beforereboot.txt](https://github.com/user-attachments/files/17640338/clinfo-beforereboot.txt)
[clinfo-afterboot.txt](https://github.com/user-attachments/files/17640337/clinfo-afterboot.txt)

Of note, I reenabled the GPU and it is folding away.

Is there someplace I can join to help with development? I've got 6 machines with 3 GPUs now going 24x7, and I'm planning on buying more.
